// amd-resolver/src/resolver.js
(function() {
  "use strict";

  var File = require('./file'),
      URL  = require('./url');

  /**
   * @constructor
   * Provides a way to build a module meta object from a module name.  The resolution
   * relies on configuration settings, which are compatible with requirejs. The created
   * module meta objects contain information such as a url that can be used for downloading
   * the corresponding file from a remote sever.
   */
  function Resolver(options) {
    this.settings = options || {};
    var baseUrl = this.settings.baseUrl || (this.settings.baseUrl = ".");

    // Make sure that if a baseUrl is provided, it ends in a slash.  This is to ensure
    // proper creation of URLs.
    if (baseUrl && baseUrl[baseUrl.length - 1] !== '/') {
      this.settings.baseUrl = baseUrl + '/';
    }
  }

  /**
   * Creates a module meta from a module name/id.
   *
   * @param {string} name - Module name/id
   * @param {string} baseUrl - base url to be used when the `name` starts with `./`, `../`, or a protocol.
   *   Otherwise the configured baseUrl is used.
   *
   * @returns {{name: string, file: File, urlArgs: string, shim: object}}
   */
  Resolver.prototype.resolve = function(name, baseUrl) {
    var i, length, file, pkg, pkgParts, pkgName, pkgTarget, shim;
    var settings = this.settings,
        urlArgs  = settings.urlArgs,
        shims    = settings.shim || {},
        packages = settings.packages || [],
        paths    = settings.paths || {},
        fileName = paths[name],
        plugins  = name.split("!");

    // The last item is the actual module name.
    name      = plugins.pop();
    pkgParts  = name.replace(/[\/\\]+/g, "/").split("/");
    pkgName   = pkgParts.shift();
    pkgTarget = pkgParts.join("/");

    // Go through the packages and figure if the module is actually configured as such.
    for (i = 0, length = packages.length; i < length; i++) {
      pkg = packages[i];

      if (pkg === pkgName) {
        fileName = pkgName + "/" + "main";
        break;
      }
      else if (pkg.name === pkgName) {
        fileName = pkg.location ? (pkg.location + "/") : "";
        fileName += pkgName + "/" + (pkgTarget || (pkg.main || "main"));
        break;
      }
    }

    if (shims.hasOwnProperty(name)) {
      shim = {
        name: shims[name].exports || shims[name].name || name,
        deps: shims[name].imports || shims[name].deps || []
      };
    }

    if (!fileName) {
       fileName = name;
    }

    // Let's assume .js extension for everything that is not defined with plugins
    if (plugins.length === 0 && /\.js$/.test(fileName) === false) {
      fileName += ".js";
    }

    baseUrl = Resolver.useBase(fileName) && baseUrl ? baseUrl : settings.baseUrl;
    file    = new File(urlArgs ? fileName + "?" + urlArgs : fileName, baseUrl);

    return {
      name: name,
      file: file, // Deprecated in favor of `url`
      url: file.url,
      shim: shim,
      plugins: plugins
    };
  };


  /**
   * Checks and returns true if name starts with `./`, `../`, or a protocol.  Otherwise returns false;
   */
  Resolver.useBase = function(name) {
    return (name[0] === '.' && (name[1] === '/' || (name[1] === '.' && name[2] === '/'))) || Resolver.hasProtocol(name);
  };


  /**
   * Quick check to determine if the name has a known protocol. Currently we only support http(s) and file.
   */
  Resolver.hasProtocol = function(name) {
    return /^(?:(https?|file)(:\/\/\/?))/g.test(name);
  };


  Resolver.File = Resolver.prototype.File = File;
  Resolver.URL  = Resolver.prototype.URL  = URL;
  module.exports = Resolver;
})();
