// amd-resolver/src/url.js
(function() {
  "use strict";

  var parser = require('url');

  /**
   * @constructor
   * Constructor for creating URL object as defined here https://developer.mozilla.org/en-US/docs/Web/API/URL
   */
  function URL(urlString, baseString) {
    var resolved = URL.parser.resolve(baseString || "", urlString);
    var url      = URL.parser.parse(resolved);

    if (url.auth) {
      var authParts = url.auth.split(":");
      url.username = authParts[0];
      url.password = authParts[1];
    }

    this.hash     = url.hash || "";
    this.host     = url.host || "";
    this.hostname = url.hostname || "";
    this.href     = url.href;
    this.origin   = url.protocol ? (url.protocol + "//" + url.host) : "";
    this.password = url.password || "";
    this.pathname = url.pathname || "";
    this.port     = url.port || "";
    this.protocol = url.protocol || "";
    this.search   = url.search || "";
    this.username = url.username || "";
  }

  URL.factory = function(url, base) {
    return new URL(url, base);
  };

  URL.parser = parser;
  module.exports = URL;
})();
